teps/0033-proxy-support.md

```diff
@@ -0,0 +1,306 @@
+enabled / configurable to benefit users/organizations to use proxy feature to communicate to
+the outside world of the platform.
+
+### User Stories (optional)
+
+User should be able to define global proxy settings on the platform and tektoncd/pipeline
+and tektoncd/triggers act based on the proxy settings defined.
+
+1. User A has http proxy setup on cluster.
+2. User B has http/https proxy setup on cluster along with certs.
+3. User C wants to run only some specific workload under proxy setup.
+
+### Notes/Constraints/Caveats (optional)
+
+1. Proposal is with constraint to have single proxy settings as an default for the environment. It does not support multi-tenancy or it will
+be using the same default proxy settings for all namespaces. But user can override it using the taskrun/pipelinerun level spec.
+
+Note: Changes required for tektoncd/triggers (yet to explore)
+
+### Risks and Mitigations
+
+None
+
+### User Experience (optional)
+
+There will be some spec changes for providing the proxy settings to pipelinerun/taskrun/pod level. But there will be no
+breaking changes and only additional spec config if user wants to leverage proxy setting. More details in design section.
+
+### Performance (optional)
+
+None
+
+## Design Details
+
+The basic idea behind the design is to propogate proxy env varibales to all the containers of pods created by controllers.
+There should be a way to define proxy environment variable which controller will read and add to the spec of all containers.
+User should be able to define these env at platform level, pipelinerun level and taskrun level.
+
+The basic flow will be
+
+1. User should be able to define env at the platform level, taskrun level and pipelinerun level. An example of taskrun look like
+
+```
+apiVersion: tekton.dev/v1alpha1
+kind: TaskRun
+metadata:
+  name: read-repo
+spec:
+  taskRef:
+    name: read-task
+  env:
+  - name: "FOO"
+    value: "bar"
+```
+
+2. The controller while scheduling the pod, make sure to propogate these env to all the containers.
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: read-repo-taskrun
+spec:
+  containers:
+  - name: prepare
+    image: docker.io/alpine@sha256:203ee936961c0f491f72ce9d3c3c67d9440cdb1d61b9783cf340baa09308ffc1
+    env:
+    - name: "FOO"
+      value: "bar"
+```
+
+User can specify a global level envs which can propagate to all the workloads. Also the global envs can be overidden
+at pipelinerun/taskrun level.
+
+### Proposed Solutions
+
+Here are the some solutions that can be a way to provide proxy support.
+
+Solution 1: 
+
+We can provide a standard configmap to define the global proxy settings. User can create a configmap with respective data
+to provide the proxy settings as configmap
+
+```
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: proxy-config
+data:
+  httpProxy: http://<username>:<pswd>@<ip>:<port>
+  httpsProxy: http://<username>:<pswd>@<ip>:<port>
+  noProxy: example.com
```

Comment:
yeah I would expect another field here for the admin to define the certs needed to communicate with a https proxy

something like a key of say `ca-bundle.crt` and the value being the pem encoded cert

Now, at least for git, it has some env's it recognizes for using a cert, or skipping tls verify (see https://stackoverflow.com/questions/11621768/how-can-i-make-git-accept-a-self-signed-certificate and https://stackoverflow.com/questions/11621768/how-can-i-make-git-accept-a-self-signed-certificate/26785963#26785963 for example) 

But that may not be the case for other binaries like the image building ones ... they typically like the certs in well known locations on the file system, with parameter overrides

Reply:
@gabemontero right, I think we may need to "define" a contract, aka "This is the set of environment variable we set and volumes we provide for certificates", and well… if the task's binaries handle things differently, it's the task responsibility to wrap/adapt :upside_down_face:. wdyt ?

Reply:
yep @vdemeester, aside from the env's, provide a volume for the cert the admin stores in the special configmap we monitor, and document how they can consume it in their task if they need it